### DISCUSSION TITLE
Analyzing instances in series before initial solve?

### DISCUSSION THREAD

**Discussion by mathgeekcoder** [ambros-gleixner/MIPcc23#11] [Q&A] [answered]
Hi!  

Is it _legal_ for us to load multiple instances in the series before we start to solve the first instance? e.g., gather statistics on changes between all instances. Obviously this would consume the time for the first instance, but I couldn't see anything in the rules that explicitly forbid this approach:

> The instances must be solved sequentially in the order specified by the input files.
> The solution file for an instance must not be modified anymore after moving on to solve the next instance.

According to the above it should technically be possible to analyze all instances and then proceed to solve them sequentially.  Is this okay, or must we load and then solve each instance sequentially? i.e., cannot load next instance until the previous is solved.

Thanks for the clarification!

**Comment by ambros-gleixner** [comment, marked as answer]
_Hi Luke,_

Thank you for this important question! You are correct that our description was vague on this point. We clarified the [rules](https://github.com/ambros-gleixner/MIPcc23/blob/main/RULES.md) now in this regard:

- The instances of one series must be solved sequentially in the order specified by the input files.
- It is not permitted to parse and analyze instances in one series ahead of time: while solving the $i$-th instance in the series, only information on the first $i$ instances may be used. The solution file for an instance must not be modified anymore after moving on to solve the next instance.

Our motivation here was that in relevant practical scenarios the data for future instances will typically not be available at this point.  It may even depend on the solution of the current instance.

However, you are of course right that information on which parts of the model may change is usually known.  If I understood you correctly, this is what you would like to extract.

We think it is fair to pass such information to your method and are currently debating how to best provide this information.  Please stand by for a few days and we will let all participants know.

Thanks again,
_Ambros_

**Reply by mathgeekcoder** [reply]
Thanks for the clarification Ambros!  That is completely fair and understandable.  

You're correct that I'd like to know what parts of the model are likely to change, and even perhaps the "range of variability" for these constants. For example, in practical scenarios some coefficients might be more volatile than others, and we might be able to exploit theoretical or empirical distributions (from historical data). 

I understand it can be tricky to provide this information, so thank you for giving it your consideration!

Cheers,
Luke

**Reply by sbolusani** [reply]
Hi Luke,

Please refer to the following changes in the latest commit regarding your query.

- The test files now contain extra rows with the names of columns and/or rows that can vary for a given component (objective function coefficients, variable bounds, constraint sides, constraint matrix) across the instances in the series.
- RULES.md contains a description of these extra rows.
- There are some minor changes to the example and evaluation scripts accordingly.

Best,
Suresh

**Reply by mathgeekcoder** [reply]
Thanks!
